File: ReverseLinkedListII.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    # @param head, a ListNode
    # @param m, an integer
    # @param n, an integer
    # @return a ListNode
    def reverseBetween(self, head, m, n):
        if not head or not head.next:
            return head
        
        dummy = ListNode(0)
        dummy.next = head
        p = dummy
        for i in xrange(m-1):
            p = p.next

        q = p.next
        for i in xrange(n-m):
            tmp = p.next
            p.next = q.next
            q.next = q.next.next
            p.next.next = tmp
            
        return dummy.next
        
so = Solution()

head = ListNode(1)
head.next = ListNode(2)
head.next.next = ListNode(3)
head.next.next.next = ListNode(4)
head.next.next.next.next = ListNode(5)

t = so.reverseBetween(head, 2, 4)

while t:
    print t.val
    t = t.next

File: CPP/Second/reverse_linked_list_ii.cc
// https://oj.leetcode.com/problems/reverse-linked-list-ii/

// Reverse a linked list from position m to n. Do it in-place and in one-pass.
// For example:
// Given 1->2->3->4->5->NULL, m = 2 and n = 4,
// return 1->4->3->2->5->NULL.
// Note:
// Given m, n satisfy the following condition:
// 1 ≤ m ≤ n ≤ length of list.

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
#include <iostream>
#include <algorithm>
#include <vector>
using namespace std;

class Solution {
public:
    ListNode *reverseBetween(ListNode *head, int m, int n) {
        ListNode dummy(-1);
        dummy.next = head;
        ListNode *p = &dummy;
		// move m-1 steps
        for (int i = 0; i < m - 1; i++)
            p = p->next;
        
        ListNode *succ = p->next;
		// do n-m-1 times switch
        for (int i = m; i < n ; i++) {
            ListNode *tmp = p->next;
            p->next = succ->next;
            succ->next = succ->next->next;
            p->next->next = tmp;
        }
        return dummy.next;
        
    }
};
int main(int argc, char *argv[])
{

    return 0;
}
